File: custom_nodes/spark/src/main/java/io/github/starter/punchline/spark/CustomSource.java
package io.github.starter.punchline.spark;

import com.fasterxml.jackson.annotation.JsonProperty;
import io.github.punchplatform.api.punchline.spark.datasets.OutputDataset;
import io.github.punchplatform.api.punchline.spark.nodes.PunchInputNode;
import java.util.LinkedList;
import java.util.List;
import java.util.stream.Collectors;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.RowFactory;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.sql.types.DataTypes;
import org.apache.spark.sql.types.Metadata;
import org.apache.spark.sql.types.StructField;
import org.apache.spark.sql.types.StructType;

/**
 * A custom source creating a single column dataset with input data
 *
 * @author Punch Team
 */
public class CustomSource extends PunchInputNode {

  @JsonProperty(value = "title")
  public String title = "a_title";

  @JsonProperty(value = "input_data")
  public List<Integer> inputData = new LinkedList<>();

  @Override
  public void execute(OutputDataset output) {
    SparkSession sparkSession = SparkSession.builder().getOrCreate();
    List<Row> rows = inputData.stream().map(RowFactory::create).collect(Collectors.toList());
    StructType schema =
        new StructType(
            new StructField[] {
              new StructField(title, DataTypes.IntegerType, false, Metadata.empty())
            });
    output.put(sparkSession.createDataFrame(rows, schema));
  }
}

File: custom_nodes/pyspark/custom_nodes/custom_source.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from api.datasets import OutputDatasets
from api.nodes import Source
from pyspark.sql import SparkSession, DataFrame

__author__ = "RT"


class CustomSource(Source):
    """Generate a Pandas DataFrame from a list of values."""

    def __init__(self, data: list) -> None:
        self.data: list = data
        if len(self.data) <= 0:
            raise ValueError("No data to inject")

    def execute(self, output_datasets: OutputDatasets) -> None:
        output: DataFrame = SparkSession.builder.getOrCreate().createDataFrame(self.data)
        output_datasets.put(output)
